hpl-2.2/man/man3/HPL_dlaswp02N.3
.TH HPL_dlaswp02N 3 "February 24, 2016" "HPL 2.2" "HPL Library Functions"
.SH NAME
HPL_dlaswp02N \- pack rows of A into columns of W.
.SH SYNOPSIS
\fB\&#include "hpl.h"\fR
 
\fB\&void\fR
\fB\&HPL_dlaswp02N(\fR
\fB\&const int\fR
\fI\&M\fR,
\fB\&const int\fR
\fI\&N\fR,
\fB\&const double *\fR
\fI\&A\fR,
\fB\&const int\fR
\fI\&LDA\fR,
\fB\&double *\fR
\fI\&W0\fR,
\fB\&double *\fR
\fI\&W\fR,
\fB\&const int\fR
\fI\&LDW\fR,
\fB\&const int *\fR
\fI\&LINDXA\fR,
\fB\&const int *\fR
\fI\&LINDXAU\fR
\fB\&);\fR
.SH DESCRIPTION
\fB\&HPL_dlaswp02N\fR
packs scattered rows of an array  A  into workspace  W.
The row offsets in A are specified by LINDXA.
.SH ARGUMENTS
.TP 8
M       (local input)           const int
On entry, M  specifies the number of rows of A that should be
copied into W. M must be at least zero.
.TP 8
N       (local input)           const int
On entry, N  specifies the length of rows of A that should be
copied into W. N must be at least zero.
.TP 8
A       (local input)           const double *
On entry, A points to an array of dimension (LDA,N). The rows
of this array specified by LINDXA should be copied into W.
.TP 8
LDA     (local input)           const int
On entry, LDA specifies the leading dimension of the array A.
LDA must be at least MAX(1,M).
.TP 8
W0      (local input/output)    double *
On exit,  W0  is  an array of size (M-1)*LDW+1, that contains
the destination offset  in U where the columns of W should be
copied.
.TP 8
W       (local output)          double *
On entry, W  is an array of size (LDW,M). On exit, W contains
the  rows LINDXA[i] for i in [0..M) of A stored  contiguously
in W(:,i).
.TP 8
LDW     (local input)           const int
On entry, LDW specifies the leading dimension of the array W.
LDW must be at least MAX(1,N+1).
.TP 8
LINDXA  (local input)           const int *
On entry, LINDXA is an array of dimension M that contains the
local row indexes of A that should be copied into W.
.TP 8
LINDXAU (local input)           const int *
On entry, LINDXAU  is an array of dimension M  that  contains
the local  row indexes of  U that should be copied into A and
replaced by the rows of W.
.SH SEE ALSO
.BR HPL_dlaswp00N \ (3),
.BR HPL_dlaswp10N \ (3),
.BR HPL_dlaswp01N \ (3),
.BR HPL_dlaswp01T \ (3),
.BR HPL_dlaswp02N \ (3),
.BR HPL_dlaswp03N \ (3),
.BR HPL_dlaswp03T \ (3),
.BR HPL_dlaswp04N \ (3),
.BR HPL_dlaswp04T \ (3),
.BR HPL_dlaswp05N \ (3),
.BR HPL_dlaswp05T \ (3),
.BR HPL_dlaswp06N \ (3),
.BR HPL_dlaswp06T \ (3).


hpl-2.2/HISTORY
==============================================================
 High Performance Computing Linpack Benchmark (HPL)
 HPL - 2.2 - February 24, 2016
==============================================================

 History

 - 09/09/00 Public release of Version 1.0

 - 09/27/00 A couple of mistakes in the  VSIPL  port have been
 corrected.  The tar file as well as the web site were updated
 on September 27th, 2000.  Note  that  these problems were not
 affecting the BLAS version of the software in any way.

 - 01/01/04 Version 1.0a
 The  MPI  process grid  numbering  scheme  is now an run-time
 option.
 The inlined assembly  timer  routine that caused the compila-
 tion to fail when using  gcc  version 3.3  and above has been
 removed from the package.
 Various building problems on the T3E have been fixed;  Thanks
 to Edward Anderson.

 - 15/12/04 Version 1.0b
 Weakness of the pseudo-random matrix generator found for pro-
 blem sizes being power of twos and larger  than 2^15;  Thanks
 to Gregory Bauer. This problem has not been fixed. It is thus
 currently recommended to  HPL  users willing to test matrices
 of size larger than 2^15 to not use power twos.

 When the matrix size is such that one needs  > 16 GB  per MPI
 rank,  the  intermediate  calculation  (mat.ld+1) * mat.nq in
 HPL_pdtest.c  ends up  overflowing  because  it is done using
 32-bit arithmetic.  This issue has been fixed by  typecasting
 to size_t; Thanks to John Baron.

 - 09/10/08 Version 2.0

 Piotr Luszczek changed to 64-bit RNG, modified files:
 -- [M] include/hpl_matgen.h
 -- [M] testing/matgen/HPL_ladd.c
 -- [M] testing/matgen/HPL_lmul.c
 -- [M] testing/matgen/HPL_rand.c
 -- [M] testing/ptest/HPL_pdinfo.c

 For a motivation for the change, see:
    Dongarra and Langou, ``The Problem with the Linpack
    Benchmark Matrix Generator'', LAWN 206, June 2008.

 -- [M] testing/ptest/HPL_pdtest.c  --

 Julien Langou changed the test for correctness from
      ||Ax-b||_oo / ( eps * ||A||_1  * N            )
      ||Ax-b||_oo / ( eps * ||A||_1  * ||x||_1      )
      ||Ax-b||_oo / ( eps * ||A||_oo * ||x||_oo * N )
 to the normwise backward error
      || r ||_oo / ( eps * ( || x ||_oo * || A ||_oo + || b ||_oo ) * N )
 See:
  Nicholas J. Higham, ``Accuracy and Stability of Numerical Algorithms'',
  Society for Industrial and Applied Mathematics, Philadelphia, PA, USA,
  Second Edition, pages = xxx+680, ISBN = 0-89871-521-0, 2002.

 Note that in our case || b ||_oo is almost for sure
 1/2, we compute it anyway.

 - 10/26/2012 Version 2.1

 Piotr Luszczek introduced exact time stamping for HPL_pdgesv():
 -- [M] dist/include/hpl_misc.h
 -- [M] dist/testing/ptest/HPL_pdtest.c

 Piotr Luszczek fixed out-of-bounds access in data spreading functions
 and exact time stamping for HPL_pdgesv():
 -- [M] dist/src/pgesv/HPL_spreadN.c
 -- [M] dist/src/pgesv/HPL_spreadT.c
 Thanks to Stephen Whalen from Cray.

 - 02/24/2016 Version 2.2

 Piotr Luszczek added continuous reporting of factorization progress
 submitted by Intel and make scripts that uses Intel software tools and
 libraries and their Apple's Mac OS X equivalents.


hpl-2.2/src/pauxil/HPL_dlaswp02N.c
/* 
 * -- High Performance Computing Linpack Benchmark (HPL)                
 *    HPL - 2.2 - February 24, 2016                          
 *    Antoine P. Petitet                                                
 *    University of Tennessee, Knoxville                                
 *    Innovative Computing Laboratory                                 
 *    (C) Copyright 2000-2008 All Rights Reserved                       
 *                                                                      
 * -- Copyright notice and Licensing terms:                             
 *                                                                      
 * Redistribution  and  use in  source and binary forms, with or without
 * modification, are  permitted provided  that the following  conditions
 * are met:                                                             
 *                                                                      
 * 1. Redistributions  of  source  code  must retain the above copyright
 * notice, this list of conditions and the following disclaimer.        
 *                                                                      
 * 2. Redistributions in binary form must reproduce  the above copyright
 * notice, this list of conditions,  and the following disclaimer in the
 * documentation and/or other materials provided with the distribution. 
 *                                                                      
 * 3. All  advertising  materials  mentioning  features  or  use of this
 * software must display the following acknowledgement:                 
 * This  product  includes  software  developed  at  the  University  of
 * Tennessee, Knoxville, Innovative Computing Laboratory.             
 *                                                                      
 * 4. The name of the  University,  the name of the  Laboratory,  or the
 * names  of  its  contributors  may  not  be used to endorse or promote
 * products  derived   from   this  software  without  specific  written
 * permission.                                                          
 *                                                                      
 * -- Disclaimer:                                                       
 *                                                                      
 * THIS  SOFTWARE  IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,  INCLUDING,  BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE UNIVERSITY
 * OR  CONTRIBUTORS  BE  LIABLE FOR ANY  DIRECT,  INDIRECT,  INCIDENTAL,
 * SPECIAL,  EXEMPLARY,  OR  CONSEQUENTIAL DAMAGES  (INCLUDING,  BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA OR PROFITS; OR BUSINESS INTERRUPTION)  HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT,  STRICT LIABILITY,  OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
 * ---------------------------------------------------------------------
 */ 
/*
 * Include files
 */
#include "hpl.h"
/*
 * Define default value for unrolling factor
 */
#ifndef HPL_LASWP02N_DEPTH
#define    HPL_LASWP02N_DEPTH       32
#define    HPL_LASWP02N_LOG2_DEPTH   5
#endif

#ifdef STDC_HEADERS
void HPL_dlaswp02N
(
   const int                        M,
   const int                        N,
   const double *                   A,
   const int                        LDA,
   double *                         W0,
   double *                         W,
   const int                        LDW,
   const int *                      LINDXA,
   const int *                      LINDXAU
)
#else
void HPL_dlaswp02N
( M, N, A, LDA, W0, W, LDW, LINDXA, LINDXAU )
   const int                        M;
   const int                        N;
   const double *                   A;
   const int                        LDA;
   double *                         W0;
   double *                         W;
   const int                        LDW;
   const int *                      LINDXA;
   const int *                      LINDXAU;
#endif
{
/* 
 * Purpose
 * =======
 *
 * HPL_dlaswp02N packs scattered rows of an array  A  into workspace  W.
 * The row offsets in A are specified by LINDXA.
 *
 * Arguments
 * =========
 *
 * M       (local input)                 const int
 *         On entry, M  specifies the number of rows of A that should be
 *         copied into W. M must be at least zero.
 *
 * N       (local input)                 const int
 *         On entry, N  specifies the length of rows of A that should be
 *         copied into W. N must be at least zero.
 *
 * A       (local input)                 const double *
 *         On entry, A points to an array of dimension (LDA,N). The rows
 *         of this array specified by LINDXA should be copied into W.
 *
 * LDA     (local input)                 const int
 *         On entry, LDA specifies the leading dimension of the array A.
 *         LDA must be at least MAX(1,M).
 *
 * W0      (local input/output)          double *
 *         On exit,  W0  is  an array of size (M-1)*LDW+1, that contains
 *         the destination offset  in U where the columns of W should be
 *         copied.
 *
 * W       (local output)                double *
 *         On entry, W  is an array of size (LDW,M). On exit, W contains
 *         the  rows LINDXA[i] for i in [0..M) of A stored  contiguously
 *         in W(:,i).
 *
 * LDW     (local input)                 const int
 *         On entry, LDW specifies the leading dimension of the array W.
 *         LDW must be at least MAX(1,N+1).
 *
 * LINDXA  (local input)                 const int *
 *         On entry, LINDXA is an array of dimension M that contains the
 *         local row indexes of A that should be copied into W.
 *
 * LINDXAU (local input)                 const int *
 *         On entry, LINDXAU  is an array of dimension M  that  contains
 *         the local  row indexes of  U that should be copied into A and
 *         replaced by the rows of W.
 *
 * ---------------------------------------------------------------------
 */ 
/*
 * .. Local Variables ..
 */
   const double               * A0 = A, * a0;
   double                     * w0;
   const int                  incA = (int)( (unsigned int)(LDA) <<
                                            HPL_LASWP02N_LOG2_DEPTH );
   int                        nr, nu;
   register int               i, j;
/* ..
 * .. Executable Statements ..
 */
   if( ( M <= 0 ) || ( N <= 0 ) ) return;

   for( i = 0; i < M; i++ ) 
      *(W0+(size_t)(i)*(size_t)(LDW)) = (double)(LINDXAU[i]);

   nr = N - ( nu = (int)( ( (unsigned int)(N) >> HPL_LASWP02N_LOG2_DEPTH ) <<
                          HPL_LASWP02N_LOG2_DEPTH ) );

   for( j = 0; j < nu;
        j += HPL_LASWP02N_DEPTH, A0 += incA, W += HPL_LASWP02N_DEPTH )
   {
      for( i = 0; i < M; i++ )
      {
         a0 = A0 + (size_t)(LINDXA[i]); w0 = W + (size_t)(i) * (size_t)(LDW);

         w0[ 0] = *a0; a0 += LDA;
#if ( HPL_LASWP02N_DEPTH >  1 )
         w0[ 1] = *a0; a0 += LDA;
#endif
#if ( HPL_LASWP02N_DEPTH >  2 )
         w0[ 2] = *a0; a0 += LDA; w0[ 3] = *a0; a0 += LDA;
#endif
#if ( HPL_LASWP02N_DEPTH >  4 )
         w0[ 4] = *a0; a0 += LDA; w0[ 5] = *a0; a0 += LDA;
         w0[ 6] = *a0; a0 += LDA; w0[ 7] = *a0; a0 += LDA;
#endif
#if ( HPL_LASWP02N_DEPTH >  8 )
         w0[ 8] = *a0; a0 += LDA; w0[ 9] = *a0; a0 += LDA;
         w0[10] = *a0; a0 += LDA; w0[11] = *a0; a0 += LDA;
         w0[12] = *a0; a0 += LDA; w0[13] = *a0; a0 += LDA;
         w0[14] = *a0; a0 += LDA; w0[15] = *a0; a0 += LDA;
#endif
#if ( HPL_LASWP02N_DEPTH > 16 )
         w0[16] = *a0; a0 += LDA; w0[17] = *a0; a0 += LDA;
         w0[18] = *a0; a0 += LDA; w0[19] = *a0; a0 += LDA;
         w0[20] = *a0; a0 += LDA; w0[21] = *a0; a0 += LDA;
         w0[22] = *a0; a0 += LDA; w0[23] = *a0; a0 += LDA;
         w0[24] = *a0; a0 += LDA; w0[25] = *a0; a0 += LDA;
         w0[26] = *a0; a0 += LDA; w0[27] = *a0; a0 += LDA;
         w0[28] = *a0; a0 += LDA; w0[29] = *a0; a0 += LDA;
         w0[30] = *a0; a0 += LDA; w0[31] = *a0; a0 += LDA;
#endif
      }
   }

   if( nr > 0 )
   {
      for( i = 0; i < M; i++ )
      {
         a0 = A0 + (size_t)(LINDXA[i]); w0 = W + (size_t)(i) * (size_t)(LDW);
         for( j = 0; j < nr; j++, a0 += LDA ) { w0[j] = *a0; }
      }
   }
/*
 * End of HPL_dlaswp02N
 */
} 
